database/src/service/read.rs

```diff
@@ -47,145 +52,211 @@ type ResponseSend = tokio::sync::oneshot::Sender<ResponseResult>;
-        // 3. Execute that function, get the result
-        // 4. Return the result via channel
-        loop {
-            // Database requests.
-            let Ok((request, response_send)) = self.receiver.recv() else {
-                // If this receive errors, it means that the channel is empty
-                // and disconnected, meaning the other side (all senders) have
-                // been dropped. This means "shutdown", and we return here to
-                // exit the thread.
-                return;
-            };
-
-            // Map [`Request`]'s to specific database functions.
-            match request {
-                ReadRequest::Example1 => self.example_handler_1(response_send),
-                ReadRequest::Example2(_x) => self.example_handler_2(response_send),
-                ReadRequest::Example3(_x) => self.example_handler_3(response_send),
-            }
+        // Create a semaphore with the same amount of
+        // permits as the amount of reader threads.
+        let semaphore = PollSemaphore::new(Arc::new(Semaphore::new(reader_count)));
+
+        // Return a handle to the pool.
+        Self {
+            pool: Arc::new(pool),
+            semaphore,
+            permit: None,
+            env: Arc::clone(env),
         }
     }
 
     /// TODO
     #[inline]
-    fn example_handler_1(&self, response_send: ResponseSend) {
-        let db_result = todo!();
-        response_send.send(db_result).unwrap();
+    pub const fn env(&self) -> &Arc<ConcreteEnv> {
+        &self.env
     }
 
     /// TODO
     #[inline]
-    fn example_handler_2(&self, response_send: ResponseSend) {
-        let db_result = todo!();
-        response_send.send(db_result).unwrap();
+    pub const fn semaphore(&self) -> &PollSemaphore {
+        &self.semaphore
     }
 
     /// TODO
     #[inline]
-    fn example_handler_3(&self, response_send: ResponseSend) {
-        let db_result = todo!();
-        response_send.send(db_result).unwrap();
+    pub const fn permit(&self) -> &Option<OwnedSemaphorePermit> {
+        &self.permit
     }
 }
+
+impl tower::Service<ReadRequest> for DatabaseReadHandle {
+    type Response = Response;
+    type Error = RuntimeError;
+    type Future = ResponseReceiver;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Check if we already have a permit.
+        if self.permit.is_some() {
+            return Poll::Ready(Ok(()));
+        }
+
+        // Acquire a permit before returning `Ready`.
+        let poll = self.semaphore.poll_acquire(cx);
+
+        match poll {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Some(permit)) => {
+                self.permit = Some(permit);
+                Poll::Ready(Ok(()))
+            }
+
+            // `self` itself owns the backing semaphore,
+            // so it can't be closed.
+            Poll::Ready(None) => unreachable!(),
+        }
```

Comment:
the `ready!` macro would probably be good here (need to import it though):
```suggestion
        let Some(permit) = ready!(self.semaphore.poll_acquire(cx)) else {
            // `self` itself owns the backing semaphore,
            // so it can't be closed.
            unreachable!()
        };
        
        self.permit = Some(permit);
        Poll::Ready(Ok(()))
        
```